third_party/proton/csrc/lib/Profiler/RoctracerProfiler.cpp

```diff
@@ -160,27 +197,118 @@ struct RoctracerProfiler::RoctracerProfilerPimpl
   auto [isRuntimeAPI, isDriverAPI] = matchKernelCbId(cid);
+
   if (!(isRuntimeAPI || isDriverAPI)) {
     return;
   }
+
   auto &profiler =
       dynamic_cast<RoctracerProfiler &>(RoctracerProfiler::instance());
-  auto &pImpl = dynamic_cast<RoctracerProfiler::RoctracerProfilerPimpl &>(
-      *profiler.pImpl);
+  auto *pImpl = dynamic_cast<RoctracerProfiler::RoctracerProfilerPimpl *>(
+      profiler.pImpl.get());
   if (domain == ACTIVITY_DOMAIN_HIP_API) {
     const hip_api_data_t *data = (const hip_api_data_t *)(callbackData);
     if (data->phase == ACTIVITY_API_PHASE_ENTER) {
       // Valid context and outermost level of the kernel launch
       auto scopeId = Scope::getNewScopeId();
       threadState.record(scopeId);
       threadState.enterOp(scopeId);
-      profiler.correlation.correlate(data->correlation_id);
+      size_t numInstances = 1;
+      if (cid == HIP_API_ID_hipGraphLaunch) {
+        pImpl->CorrIdToIsHipGraph[data->correlation_id] = true;
+        hipGraphExec_t GraphExec = data->args.hipGraphLaunch.graphExec;
+        numInstances = std::numeric_limits<size_t>::max();
+        bool findGraph = false;
+        if (pImpl->GraphExecToGraph.contain(GraphExec)) {
+          hipGraph_t Graph = pImpl->GraphExecToGraph[GraphExec];
+          if (pImpl->GraphToNumInstances.contain(Graph)) {
+            numInstances = pImpl->GraphToNumInstances[Graph];
+            findGraph = true;
+          }
+        }
+        if (!findGraph)
+          std::cerr
+              << "[PROTON] Cannot find graph and it may cause a memory leak."
+                 "To avoid this problem, please start profiling before the "
+                 "graph is created."
+              << std::endl;
+      }
+      profiler.correlation.correlate(data->correlation_id, numInstances);
     } else if (data->phase == ACTIVITY_API_PHASE_EXIT) {
+      switch (cid) {
+      case HIP_API_ID_hipStreamBeginCapture: {
+        hipStream_t Stream = data->args.hipStreamBeginCapture.stream;
+        pImpl->StreamToCaptureCount[Stream] = 0;
+        pImpl->StreamToCapture[Stream] = true;
+        break;
+      }
+      case HIP_API_ID_hipStreamEndCapture: {
+        hipGraph_t Graph = *(data->args.hipStreamEndCapture.pGraph);
+        hipStream_t Stream = data->args.hipStreamEndCapture.stream;
+        // How many times did we capture a kernel launch for this steam
+        uint32_t StreamCaptureCount = pImpl->StreamToCaptureCount[Stream];
+        pImpl->GraphToNumInstances[Graph] = StreamCaptureCount;
+        pImpl->StreamToCapture.erase(Stream);
+      }
+      case HIP_API_ID_hipLaunchKernel: {
+        hipStream_t Stream = data->args.hipLaunchKernel.stream;
+        if (pImpl->StreamToCapture.contain(Stream))
+          pImpl->StreamToCaptureCount[Stream]++;
+        break;
+      }
+      case HIP_API_ID_hipExtLaunchKernel: {
+        hipStream_t Stream = data->args.hipExtLaunchKernel.stream;
```

Comment:
It is specifically for if you want to launch a kernel with a function pointer as the kernel definition instead of a traditional direct call. But in more recent Rocm builds you can you can pass a function pointer directly to hipLaunchKernel. I believe there are some PyTorch builds that rely on older Rocm version so left it in just in case.